ArduCopter/mode_systemid.cpp

```diff
@@ -0,0 +1,361 @@
+                    break;
+                case MIX_ROLL:
+                    attitude_control->actuator_roll_sysid(waveformSample);
+                    break;
+                case MIX_PITCH:
+                    attitude_control->actuator_pitch_sysid(waveformSample);
+                    break;
+                case MIX_YAW:
+                    attitude_control->actuator_yaw_sysid(waveformSample);
+                    break;
+                case MIX_THROTTLE:
+                    pilot_throttle_scaled += waveformSample;
+                    break;
+            }
+            break;
+    }
+
+    // call attitude controller
+    attitude_control->input_euler_angle_roll_pitch_euler_rate_yaw(target_roll, target_pitch, target_yaw_rate);
+
+    // body-frame rate controller is run directly from 100hz loop
+
+    // output pilot's throttle
+    if (copter.is_tradheli()) {
+        attitude_control->set_throttle_out(pilot_throttle_scaled, false, g.throttle_filt);
+    } else {
+        attitude_control->set_throttle_out(pilot_throttle_scaled, true, g.throttle_filt);
+    }
+
+    if(log_subsample <= 0){
+        log_data();
+        if(copter.should_log(MASK_LOG_ATTITUDE_FAST) && copter.should_log(MASK_LOG_ATTITUDE_MED)) {
+            log_subsample = 1;
+        } else if(copter.should_log(MASK_LOG_ATTITUDE_FAST)){
+            log_subsample = 2;
+        } else if(copter.should_log(MASK_LOG_ATTITUDE_MED)){
+            log_subsample = 4;
+        } else {
+            log_subsample = 8;
+        }
+    }
+    log_subsample -= 1;
+
+    log_data();
+}
+
+// init_test - initialises the test
+void ModeSystemId::log_data()
+{
+    int8_t index = copter.EKF2.getPrimaryCoreIMUIndex();
+    if(index < 0){
+        index = 0;
+    }
+
+    Vector3f delta_angle;
+    copter.ins.get_delta_angle(index, delta_angle);
+    float delta_angle_dt = copter.ins.get_delta_angle_dt(index);
+
+    Vector3f delta_velocity;
+    copter.ins.get_delta_velocity(index, delta_velocity);
+    float delta_velocity_dt = copter.ins.get_delta_velocity_dt(index);
+
+    if(is_positive(delta_angle_dt) && is_positive(delta_velocity_dt)){
+        AP::logger().Write(
+            "SIDI",
+            "TimeUS,Time,Targ,F,Gx,Gy,Gz,Ax,Ay,Az",
+            "ss-zkkkooo",
+            "F---------",
+            "Qfffffffff",
+            AP_HAL::micros64(),
+            (double)waveformTime,
+            (double)waveformSample,
+            (double)waveformFreqRads / (2 * M_PI),
+            (double)(degrees(delta_angle.x / delta_angle_dt)),
+            (double)(degrees(delta_angle.y / delta_angle_dt)),
+            (double)(degrees(delta_angle.z / delta_angle_dt)),
+            (double)(delta_velocity.x / delta_velocity_dt),
+            (double)(delta_velocity.y / delta_velocity_dt),
+            (double)(delta_velocity.z / delta_velocity_dt));
+    }
+    // Full rate logging of attitude, rate and pid loops
+    copter.Log_Write_Attitude();
```

Comment:
The attitude logging can already run at 400hz.  Is there any reason why we're not using the fourhundred_hz_logging() function to do the logging?  The advantage of using the existing fourhundred_hz_logging() method is that we don't have to try as hard to ensure we don't do it twice and also it means the logging is done outside the fast loop.  It may miss some logs but the main loop will run more regularly.